src/app/util/MatterCallbacks.h

```diff
@@ -14,32 +13,57 @@
  *    See the License for the specific language governing permissions and
  *    limitations under the License.
  */
-
-// THIS FILE IS GENERATED BY ZAP
-
 #pragma once
 
 #include <access/SubjectDescriptor.h>
 #include <app/ConcreteAttributePath.h>
 #include <app/ConcreteCommandPath.h>
 
-void MatterPreAttributeReadCallback(const chip::app::ConcreteAttributePath & attributePath);
-void MatterPostAttributeReadCallback(const chip::app::ConcreteAttributePath & attributePath);
-void MatterPreAttributeWriteCallback(const chip::app::ConcreteAttributePath & attributePath);
-void MatterPostAttributeWriteCallback(const chip::app::ConcreteAttributePath & attributePath);
+namespace chip {
 
-/** @brief Matter Pre Command Received
- *
- * This callback is called once the message has been determined to be a command, and
- * before the command is dispatched to the receiver.
- */
-CHIP_ERROR MatterPreCommandReceivedCallback(const chip::app::ConcreteCommandPath & commandPath,
-                                            const chip::Access::SubjectDescriptor & subjectDescriptor);
+/// Allows for application hooks for processing attribute and command operations
+class DataModelCallbacks
+{
+public:
+    enum class OperationType
+    {
+        Read,
+        Write
+    };
 
-/** @brief Matter Post Command Received
- *
- * This callback is called once the message has been determined to be a command, but
- * after it being dispatched to the receiver.
- */
-void MatterPostCommandReceivedCallback(const chip::app::ConcreteCommandPath & commandPath,
-                                       const chip::Access::SubjectDescriptor & subjectDescriptor);
+    enum class OperationOrder
+    {
+        Pre,
+        Post
+    };
+
+    virtual ~DataModelCallbacks() = default;
+
+    /// This callback is called on attribute operations:
+    /// - for reads and writes
+    /// - both befor and after attribute read/writes
+    ///
+    /// NOTE: PostRead is only called on read success.
+    virtual void AttributeOperation(OperationType operation, OperationOrder order, const chip::app::ConcreteAttributePath & path) {}
+
+    /// This callback is called once the message has been determined to be a command, and
```

Comment:
This comment doesn't make sense for messages with more than one command path in them.  What are the actual semantics here?

Reply:
I copied over what already existed in comments in src/app/util/MatterCallbacks.h

I will take a look, however I am somewhat weary of increasing the scope of the PR from a mechanical 'move things around' to being responsible for more correctness bits. I believe part of the code is already dubious: why does write call the post even on error (but error code is not reported to it) yet read post is not called on error.

Overall I think cleanup should follow the moving around as a separate work item.

Reply:
Updated the comment: the logic seems to be related to CommandDataIB so it will be called several times for batch invoke. Hoping the new comment is clearer.